Move decompositions and helpers for jvp from functorch into core

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #84151
* __->__ #84358

This refactor shouldn't change any behavior. At this point functorch still relies on the mechanism in DynamicLayerFront; we just moved some parts of it into core.

soulitzer committed:
More decompositions for jvp from functorch into core

[ghstack-poisoned]

soulitzer committed:
Update on "Move decompositions for jvp from functorch into core"




[ghstack-poisoned]

soulitzer commented on torch/_decomp/decompositions_for_jvp.py:
```diff
@@ -1,17 +1,31 @@
+#
+# TODO: what if jit decompositions exists, should we just use it?
+#       or do we want to have an explicit white list like functorch had
+#       using special JVP_DECOMP DynamicLayerFront kernel
```
This probably belongs in the next PR of the stack, but happy to keep it here too

samdow replied:
Yeah either's fine to me

soulitzer committed:
Update on "Move decompositions and helpers for jvp from functorch into core"


This refactor shouldn't change any behavior. At this point functorch still relies on the mechanism in DynamicLayerFront; we just moved some parts of it into core.

[ghstack-poisoned]

soulitzer commented on torch/_decomp/decompositions_for_jvp.py:
```diff
@@ -33,6 +48,39 @@ def register_decomposition_for_jvp(fn):
+
+# The only decompositions here are temporary or hacks for the purposes of jvp
+
+# TODO: do these also belong here?
```
I rename _register_jit_decomposition to _register_jit_decomposition_for_jvp, assuming that the only place it is used is for jvp - not completely sure if that is true

samdow replied:
Yep! This makes sense to me. Backstory is it is only currently used for jvp, but since I think it took a couple people to build this, we didn't update the name.

samdow commented on torch/_decomp/decompositions_for_jvp.py:
```diff
@@ -145,7 +195,9 @@ def prod(x: List[int]):
-@register_decomposition_for_jvp(aten.native_batch_norm_backward)  # @register_decomposition_for_jvp after in core
+@register_decomposition_for_jvp(
+    aten.native_batch_norm_backward
+)  # @register_decomposition_for_jvp after in core
```
Super nit but since we're in here, could we remove this comment? This should have been removed earlier by us

soulitzer committed:
Update on "Move decompositions and helpers for jvp from functorch into core"


This refactor shouldn't change any behavior. At this point functorch still relies on the mechanism in DynamicLayerFront; we just moved some parts of it into core.

[ghstack-poisoned]

soulitzer committed:
Update on "Move decompositions and helpers for jvp from functorch into core"


This refactor shouldn't change any behavior. At this point functorch still relies on the mechanism in DynamicLayerFront; we just moved some parts of it into core.

[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g

malfet commented:
@pytorchbot revert -m "Broke lint" -c landrace
